new: initiate gaia spec for enforcer auto upgrade

t00f opened the pull request:
PR for the enforcer auto upgrade feature (https://redlock.atlassian.net/browse/CNS-33)

> Fixes https://redlock.atlassian.net/browse/CNS-268

t00f commented on specs/enforcer.spec:
```diff
@@ -270,12 +270,27 @@ attributes:
     transient: true
 
-  - name: updateAvailable
-    description: If `true`, the enforcer version is outdated and should be updated.
```
@primalmotion I can't remember why we have this attribute and if it's still used or not?

primalmotion replied:
It's not. Can be removed

brianonn commented on specs/enforcer.spec:
```diff
@@ -270,12 +270,27 @@ attributes:
+    orderable: true
+
+  - name: upgradeTargetVersion
+    description: Defines the next version the enforcer will be upgraded to.
```
we also talked about downgrades / rollbacks. and in that context using `upgradeTargetVersion` might not make semantic sense.   How about `requestedTargetVersion` ? 
```
description: Defines the next version the enforcer will be moved to
```

primalmotion replied:
This property is read only. No need for validation here

brianonn commented on specs/enforcer.spec:
```diff
@@ -270,12 +270,27 @@ attributes:
+    - None
+    - UpgradeRequested
+    - Upgrading
+    default_value: None
```
maybe just `None`, `Requested` and `Upgrading` but then, we also talked about upgrades and rollbacks, so I want to be able to cover that.  Meaning, maybe lets use simply `None`, `Requested` and `InProgress` states.

brianonn replied:
if we call `upgradeStatus` as `targetVersionStatus` (see line 273)  then these three states can be `None`, `ChangeRequested` and `InProgress`

brianonn commented on specs/enforcer.spec:
```diff
@@ -270,12 +270,27 @@ attributes:
-  - name: updateAvailable
-    description: If `true`, the enforcer version is outdated and should be updated.
-    type: boolean
+  - name: upgradeStatus
```
since we wanted to consider upgrades and rollbacks as operationally similar things, maybe this property can be called `targetStatus` to follow along with the idea I presented below, `requestedTargetVersion`

primalmotion replied:
Fine by me

primalmotion replied:
This property is read only

brianonn commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
+    exposed: true
+    read_only: true
+    autogenerated: true
+
```
who sets the recursive flag, and who consumes it? I don't need it at the enforcer object but maybe it's needed by the UI to tell the backend something?  I just want to understand how it's used 

Also, I want to remove the word 'upgrade' as a prefix on these properties, since we want to try to consider an upgrade and a rollback operationally the same thing i.e. it's just a version change at the enforcer.  

So I've been trying to think of these as simply a `targetVersion` or `requestedVersion` and use that as a prefix instead of `upgrade...`

t00f replied:
The recursive will be useful in the case we want the "upgrade" to happen in a namespace recursively. 

If you receive an enforcerRefresh with `namespace: /cns` and `upgradeRecursive: true`, it means all enforcers in /cns and below should be "upgraded".

brianonn commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
     default_value: Counters
     omit_empty: true
 
```
Marcus and I had also tossed about the idea of using a `refreshType` (or can be simply `type` ) field on the enforcerrefresh spec. so that we can differentiate between a debug type refresh and a `change version` type of refresh.   I.e. we thought maybe we needed an enum here for this purpose.   

Granted, we can assume one type or the other based on what fields are actually populated in the refresh we see, but this is not always the right way to do it.

primalmotion replied:
Yeah. We should have a enum for type

brianonn commented on specs/namespace.spec:
```diff
@@ -125,6 +125,14 @@ attributes:
+    validations:
+    - $semver
+
   - name: name
```
@t00f Indeed at one point we talked about a `minimumEnforcerVersion` but I believe it was resolved to be the `requestedEnforcerVersion` instead.  i.e. we don't want to imply that there is any policy level enforcement of this minimum.  

Antoine @primalmotion can add his thoughts here as well, but I think this `minimumEnforcerVersion` should be called `desiredEnforcerVersion` or even `bootstrapEnforcerVersion` to indicate that it is used when new enforcers startup and register into the namespace. 

Actually, the name `bootstrapEnforcerVersion` has some appeal since it corresponds to the new install / upgrade flow that we are designing, where the `bootstrap` code starts up in an AMI and launches the `core` /  `updater` process which  then downloads and bootstraps the enforcer in this namespace, using the version in this field, or walking up the namespace hierarchy to the parents until it finds one (like CA), or when not present and not found on the parent paths then it will use the latest enforcer version.

brianonn replied:
could also be `initialEnforcerVersion` or `defaultEnforcerVersion`

t00f reviewed (commented):
These are valid comments, thanks Brian.

I am suggesting to use `migrate` instead of `upgrade`:

* `migrationVersion` : Defines the next version the enforcer will be moved to
* `migrationStatus` (None, Requested, Migrating): Define the migration status
* `migrationSelector`: Define which enforcers are targeted by the migration
* `migrationRecursive`: Define if the migration should be recursive in the given namespace
* `defaultEnforcerVersion` makes more sense to me

Note: The UI will talk about upgrade, and not migration, correct?
What do you think?

primalmotion commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
+    type: external
+    exposed: true
+    subtype: '[][]string'
+    read_only: true
```
This property is not read only

primalmotion commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
 
-  - name: debugID
-    description: Can be used to correlate with a DebugBundle.
+  - name: upgradeRecursive
```
Not read only

primalmotion commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
+    description: Indicates if the upgrade should be done recursively in all namespaces.
+    type: boolean
+    exposed: true
+    read_only: true
```
Not auto generated

primalmotion commented on specs/enforcerrefresh.spec:
```diff
@@ -36,28 +36,57 @@ attributes:
+    description: Defines the version to upgrade the enforcers.
     type: string
     exposed: true
+    read_only: true
```
Not read only and not autogenerated

t00f commented:
@brianonn @primalmotion I addressed all your comments in the latest changes. We are now talking about a "migration" of the enforcer instead of an "upgrade"

t00f commented:
#### Notes

* Enforcer -> Check version `migrationAvailableVersion` 
* UI sends an enforcerRefresh  Enforcer Push -> Status = Running
  * a) Success = `migrationAvailableVersion`=null + Status = None
  * b) Failure = keep `migrationAvailableVersion` + Status = Failed + Send an alarm + event log.
* /!\ In the processor of enforcerRefresh, If no parentID, requires a selector
* Avoid the semver dependency

brianonn commented:
I commented in Slack already on Monday, but I'll also just add my approval here.
It looks good, let's go with this and I'll use this PR with the enforcer to develop it out in the next sprints.

brianonn commented:
> These are valid comments, thanks Brian.
> 
> I am suggesting to use `migrate` instead of `upgrade`:
> 
> * `migrationVersion` : Defines the next version the enforcer will be moved to
> * `migrationStatus` (None, Requested, Migrating): Define the migration status
> * `migrationSelector`: Define which enforcers are targeted by the migration
> * `migrationRecursive`: Define if the migration should be recursive in the given namespace
> * `defaultEnforcerVersion` makes more sense to me
> 
> Note: The UI will talk about upgrade, and not migration, correct?
> What do you think?

I suppose the UI can also be smart about it and present [ Upgrade ] or [ Downgrade ] buttons if it has all the knowledge about the semantic ordering of the currentVersion and the `nextAvailableVersion`  

@primalmotion @t00f With `nextAvailableVersion` , do you think we also need a `desiredVersion` or `requestedVersion` ?
I'm a little confused on the intended use for `nextAvailableVersion`.  That name makes it seem like it's informational only, but then it's also to be used to migrate to? so it's actually the `requestedVersion`... do we need both ?

t00f commented:
@brianonn,

`nextAvailableVersion` is to provide which version the enforcer should be **updated**. If a user is planning a rollback, the `migrationVersion` is the one that is sent in the enforcerRefresh notification

brianonn commented:
> @brianonn,
> 
> `nextAvailableVersion` is to provide which version the enforcer should be **updated**. If a user is planning a rollback, the `migrationVersion` is the one that is sent in the enforcerRefresh notification

ok thanks for the clarification! @t00f

ericrpowers commented on custom_validations.go:
```diff
@@ -1107,6 +1108,17 @@ func ValidateUIParameters(p *UIParameter) error {
 }
 
+// ValidateSemVer validates a semantic version.
+func ValidateSemVer(attribute, data string) error {
+
+	_, err := semver.Parse(strings.TrimPrefix(data, "v"))
+	if err != nil {
+		return makeValidationError(attribute, fmt.Sprintf("invalid semver %s: %s", data, err))
+	}
+
+	return nil
+}
```
We probably need to handle empty string here as `defaultEnforcerVersion` is not a required value:
```
	if data == "" {
		return nil
	}
```

t00f commented:
Resync'ed with latest changes from Amir :)
